content/actions/automating-builds-and-tests/building-and-testing-ruby.md

```diff
@@ -87,13 +87,13 @@ Alternatively, you can check a `.ruby-version` file  into the root of your repos
 
 ## Testing with multiple versions of Ruby
 
-You can add a matrix strategy to run your workflow with more than one version of Ruby. For example, you can test your code against the latest patch releases of versions 2.7, 2.6, and 2.5. The 'x' is a wildcard character that matches the latest patch release available for a version.
+You can add a matrix strategy to run your workflow with more than one version of Ruby. For example, you can test your code against the latest patch releases of versions 3.0, 2.7, and 2.6.  YAML parsers will truncate trailing zeros from numeric values, and often convert floats to integers.  So, an unquoted 3.0 is converted to an integer, so it will match any minor version, as in 3.0.3, 3.1.0, etc.  Likewise, 2.10 needs to be quoted, otherwise it will be interpreted as 2.1.
```

Comment:
```suggestion
You can add a matrix strategy to run your workflow with more than one version of Ruby. For example, you can test your code against the latest patch releases of versions 3.0, 2.7, and 2.6.  The YAML parser used by GitHub Actions will truncate trailing zeros from numeric values, and often convert floats to integers.  So, an unquoted 3.0 is converted to an integer, so it will match any minor version, as in 3.0.3, 3.1.0, etc.  Likewise, 2.10 needs to be quoted, otherwise it will be interpreted as 2.1.
```

It is good that we document this gotcha, but it sounds like it is a bug that should be addressed: https://github.com/actions/runner/issues/849#issuecomment-984495423

Reply:
IDK.  The text I wrote was meant to not be specific, as most coders would prefer [3.0, 2.7, 2.6] to ['3.0', '2.7', '2.6'].  Each item is really an array match to a version string (split into an array), which isn't a number (ie, major.minor.patch, or 3.0.1, worse with alphas, betas, rc's, etc).  Hence, I wanted to make readers aware of it without implying a yaml parser is working incorrectly.  Not the best explanation, but...

**off-topic:** For people who don't code Ruby, it's a non-typed language, but 1/3 = 0 and 1/3.0 = 0.333...
Hence, Ruby considers 3 to be an int, and 3.0 to be a float.  So, Ruby's default yaml parser considers `3` to be an int, and `3.0` to be a float.

Reply:
Coders may prefer `[3.0, 2.7, 2.6]` for simplicity, but those aren't numbers, they're strings and they should be quoted.  Otherwise you will run into exactly the problem that you describe where `2.10` needs to be quoted to disambiguate from `2.1`.  I would illustrate the happy path in the doc and then explain _why_ you've quoted _all_ the version numbers instead of trying to tell people when they need to quote.

Reply:
@ethomson

Thanks.  I don't know what's best in this doc, but there are a lot of workflows in the wild that only quote what is needed, ie '3.0'.  So, I wanted some background on why...  Also, see #15240

Reply:
That makes sense @MSP-Greg.  I think that it's safe to say that mine is an unpopular opinion, so I'd like to give some time to give others a chance to provide feedback, but I think that we should encourage users to quote version numbers and ensure that the starter workflows follow our guidance here.

Reply:
> Coders may prefer [3.0, 2.7, 2.6] for simplicity, but those aren't numbers, they're strings and they should be quoted. 

I agree with this. The semantic version isn't actually a number, it just works in some cases to treat it as such. Although people often leave out the quotes, we try to craft examples in the docs that illustrate best practices and that will work if users copy-paste-edit without being aware of nuances like this.

@dentarg opened a PR that does this: https://github.com/github/docs/pull/15240